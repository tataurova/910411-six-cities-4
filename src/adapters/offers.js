export const getOffer = (data) => {
  return {
    id: data.id,
    title: data.title,
    city: data.city.name,
    cityCoordinates: [data.city.location.latitude, data.city.location.longitude],
    cityZoom: data.city.location.zoom,
    coordinates: [data.location.latitude, data.location.longitude],
    type: data.type,
    price: data.price,
    rating: data.rating,
    premium: data.is_premium,
    favorite: data.is_favorite,
    photo: data.preview_image,
    bedrooms: data.bedrooms,
    adults: data.max_adults,
    additional: data.goods,
    details: data.description,
    owner: {
      photo: data.host.avatar_url,
      name: data.host.name,
      isSuper: data.host.is_pro,
    },
    photos: data.images,
    reviews: [ // todo reviews from server
      {
        name: `Max`,
        photo: `img/avatar-max.jpg`,
        rating: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        The building is green and from 18th century.`,
        date: `April 2019`,
      },
    ],
  };
};
