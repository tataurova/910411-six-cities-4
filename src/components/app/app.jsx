import React, {PureComponent} from 'react';
import Main from '../main/main.jsx';
import PlaceFullCard from "../place-full-card/place-full-card.jsx";
import PropTypes from 'prop-types';
import {Switch, Route, BrowserRouter} from "react-router-dom";
import {placeCardType, mapSettingsType} from "../../../types.js";

class App extends PureComponent {
  constructor(props) {
    super(props);
    this.handleCardHeaderClick = this.handleCardHeaderClick.bind(this);
    this.state = {
      page: 0,
    };
  }

  handleCardHeaderClick(id) {
    this.setState({page: id});
  }

  _renderApp() {
    return <Main
      offers = {this.props.offers}
      placeCount = {this.props.placeCount}
      mapSettings = {this.props.mapSettings}
      onPlaceCardHeaderClick = {this.handleCardHeaderClick}
    />;
  }

  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
          <Route exact path="/offer/:id" render={(props) =>
            <PlaceFullCard
              offers = {this.props.offers}
              mapSettings = {this.props.mapSettings}
              onPlaceCardHeaderClick = {this.handleCardHeaderClick}
              {...props.match.params}
            />
          }
          />
        </Switch>
      </BrowserRouter>
    );
  }
}

export default App;

App.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.shape(placeCardType)).isRequired,
  mapSettings: PropTypes.shape(mapSettingsType).isRequired,
  placeCount: PropTypes.number.isRequired,
};

