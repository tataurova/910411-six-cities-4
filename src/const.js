export const PLACES_COUNT = 312;

export const MapSettings = {
  city: [52.38333, 4.9],
  icon: {
    iconUrl: `img/pin.svg`,
    iconSize: [30, 30],
  },
  zoom: 12,
};
