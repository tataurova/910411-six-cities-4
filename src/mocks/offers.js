export default [
  {
    id: 1,
    title: `Canal View Prinsengracht`,
    city: `Cologne`,
    cityCoordinates: [50.938361, 6.959974],
    cityZoom: 13,
    coordinates: [50.960361, 6.967974],
    type: `Apartment`,
    price: 120,
    rating: 4,
    premium: true,
    photo: `img/apartment-01.jpg`,
    bedrooms: 3,
    adults: 6,
    additional: [`Wifi`, `Heating`, `Kitchen`],
    details: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
     The building is green and from 18th century. An independent House, strategically located between Rembrand
     Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    owner: {
      photo: `img/avatar-angelina.jpg`,
      name: `Angelina`,
      isSuper: true,
    },
    photos: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/room.jpg`
    ],
    reviews: [
      {
        name: `Max`,
        photo: `img/avatar-max.jpg`,
        rating: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        The building is green and from 18th century.`,
        date: `April 2019`,
      },
    ],
  }, {
    id: 2,
    title: `Wood and stone place`,
    city: `Cologne`,
    cityCoordinates: [50.938361, 6.959974],
    cityZoom: 13,
    coordinates: [50.913361, 6.9509739999999995],
    type: `Private room`,
    price: 80,
    rating: 4,
    premium: false,
    photo: `img/room.jpg`,
    bedrooms: 2,
    adults: 4,
    additional: [`Wifi`, `Kitchen`, `Cable TV`, `Washing machine`],
    details: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
     The building is green and from 18th century. An independent House, strategically located between Rembrand
     Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    owner: {
      photo: `img/avatar-max.jpg`,
      name: `Max`,
      isSuper: false,
    },
    photos: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/room.jpg`
    ],
    reviews: [
      {
        name: `Max`,
        photo: `img/avatar-max.jpg`,
        rating: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        The building is green and from 18th century.`,
        date: `April 2019`,
      },
    ],
  }
];

/* eslint-disable */
export const serverOffers = [
  {
    id: 1,
    title: `Canal View Prinsengracht`,
    city: {
      name: `Cologne`,
      location: {
        latitude: 50.938361,
        longitude: 6.959974,
        zoom: 13,
      },
    },
    location: {
      latitude: 50.960361,
      longitude: 6.967974,
    },
    type: `Apartment`,
    price: 120,
    rating: 4,
    is_premium: true,
    preview_image: `img/apartment-01.jpg`,
    bedrooms: 3,
    max_adults: 6,
    goods: [`Wifi`, `Heating`, `Kitchen`],
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
     The building is green and from 18th century. An independent House, strategically located between Rembrand
     Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    host: {
      id: 1,
      avatar_url: `img/avatar-angelina.jpg`,
      name: `Angelina`,
      is_pro: true,
    },
    images: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/room.jpg`
    ],
    reviews: [
      {
        name: `Max`,
        photo: `img/avatar-max.jpg`,
        rating: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        The building is green and from 18th century.`,
        date: `April 2019`,
      },
    ],
  }, {
    id: 2,
    title: `Wood and stone place`,
    city: {
      name: `Cologne`,
      location: {
        latitude: 50.938361,
        longitude: 6.959974,
        zoom: 13,
      },
    },
    location: {
      latitude: 50.913361,
      longitude: 6.9509739999999995,
    },
    type: `Private room`,
    price: 80,
    rating: 4,
    is_premium: false,
    preview_image: `img/room.jpg`,
    bedrooms: 2,
    max_adults: 4,
    goods: [`Wifi`, `Kitchen`, `Cable TV`, `Washing machine`],
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
     The building is green and from 18th century. An independent House, strategically located between Rembrand
     Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    host: {
      id: 2,
      avatar_url: `img/avatar-max.jpg`,
      name: `Max`,
      is_pro: false,
    },
    images: [
      `img/room.jpg`,
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
      `img/studio-01.jpg`,
      `img/room.jpg`
    ],
    reviews: [
      {
        name: `Max`,
        photo: `img/avatar-max.jpg`,
        rating: 4,
        text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
        The building is green and from 18th century.`,
        date: `April 2019`,
      },
    ],
  }
];
