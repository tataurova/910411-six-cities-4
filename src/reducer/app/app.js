import {extend} from "../../utils/common.js";

const initialState = {
  cities: [],
  city: ``,
};

const ActionType = {
  CHANGE_CITY: `CHANGE_CITY`,
  SET_CITIES: `SET_CITIES`,
};

export const ActionCreator = {
  changeCity: (city) => ({
    type: ActionType.CHANGE_CITY,
    payload: city,
  }),
  setCities: (offers) => ({
    type: ActionType.SET_CITIES,
    payload: [...new Set(offers.map((offer) => offer.city))],
  }),
};


const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return extend(state, {
        city: action.payload,
      });
    case ActionType.SET_CITIES:
      return extend(state, {
        cities: action.payload,
        city: action.payload[0],
      });
  }

  return state;
};

export {reducer, ActionType};
