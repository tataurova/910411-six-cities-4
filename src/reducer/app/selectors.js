import {createSelector} from "reselect";
import NameSpace from "../name-space.js";

export const getOffers = (state) => {
  return state[NameSpace.DATA].offers;
};

export const getActiveCity = (state) => {
  return state[NameSpace.APP].city;
};

export const getMemoizedCityOffers = createSelector(
    getOffers,
    getActiveCity,
    (offers, city) => {
      return offers.filter((offer) => offer.city === city);
    }
);
