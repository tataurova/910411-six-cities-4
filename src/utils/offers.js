export const getOfferInfo = (offers, offerId) => {
  const idNumber = Number(offerId);
  return {
    offer: offers.filter((item) => item.id === idNumber)[0],
    nearOffers: offers.filter((item) => item.id !== idNumber),
  };
  // for (let key in offers) {
  //   if (offers.hasOwnProperty(key)) {
  //     const offer = offers[key].filter((item) => item.id === idNumber)[0];
  //     if (offer) {
  //       return {
  //         offer,
  //         nearOffers: offers[key].filter((item) => item.id !== offer.id),
  //       };
  //     }
  //   }
  // }
  // return {};
};
