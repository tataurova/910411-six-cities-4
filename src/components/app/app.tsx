import React from 'react';
import MainPage from '../main-page/main-page.js';
import PlaceFullCard from "../place-full-card/place-full-card.js";
import PropTypes from 'prop-types';
import {Switch, Route, Router, Redirect} from "react-router-dom";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer/app/app.js";
import {placeCardType, reviewType} from "../../../types.js";
import {findOffer} from "../../utils/offers.js";
import {getMemoizedCityOffers} from "../../reducer/app/selectors.js";
import {getMemoizedOffers} from "../../reducer/data/selectors.js";
import NameSpace from "../../reducer/name-space.js";
import Login from "../login/login.js";
import withAuthentication from "../../hocs/with-authentication/with-authentication.jsx";
import {Operation as UserOperation} from "../../reducer/user/user.js";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import {AppRoute, AuthorizationStatus} from "../../const.js";
import history from "../../history.js";
import Favorites from "../favorites/favorites.js";
import PrivateRoute from "../private-route/private-route.js";
import {CardType} from "../../const.js";
import NotFound from "../not-found/not-found.js";

const LoginWithAuthentication = withAuthentication(Login);

class App extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {
      offers,
      login,
      authorizationStatus,
      user,
      sendComment,
      isFetching,
      error,
      favoriteOffers,
      loadFavoriteOffers,
      loadReviews,
      reviews,
      loadNearbyOffers,
      nearbyOffers,
      onBookmarkButtonCLick,
    } = this.props;
    return (
      <Router history = {history}>
        <Switch>
          <Route exact path={AppRoute.MAIN} render={() => {
            return <MainPage
              {...this.props}
            />;
          }}>
          </Route>
          <Route exact path={`${AppRoute.PLACE_FULL_CARD}/:id`} render={(props) => {
            const offer = findOffer(offers, props.match.params.id);
            if (offer) {
              return <PlaceFullCard
                offer = {findOffer(offers, props.match.params.id)}
                {...props.match.params}
                authorizationStatus = {authorizationStatus}
                user = {user}
                onSubmitForm = {sendComment}
                isFetching = {isFetching}
                error = {error}
                loadReviews = {loadReviews}
                reviews = {reviews}
                loadNearbyOffers = {loadNearbyOffers}
                nearbyOffers = {nearbyOffers}
                onBookmarkButtonCLick = {onBookmarkButtonCLick}
              />;
            } else {
              return <NotFound
                authorizationStatus = {authorizationStatus}
                user = {user}
              />;
            }
          }
          }
          />
          <Route exact path={AppRoute.LOGIN} render={() => {
            switch (authorizationStatus) {
              case AuthorizationStatus.AUTH:
                return <Redirect to={AppRoute.MAIN} />;
              case AuthorizationStatus.NO_AUTH:
                return <LoginWithAuthentication
                  onSubmitForm = {login}
                  authorizationStatus = {authorizationStatus}
                  user = {user}
                  error = {error}
                />;
              default:
                throw error(`Unknown AuthorizationStatus ${authorizationStatus}`);
            }
          }
          }
          />
          <PrivateRoute
            exact
            path={AppRoute.FAVORITES}
            render={() => {
              return (
                <Favorites
                  cardType = {CardType.FAVORITE}
                  favoriteOffers = {favoriteOffers}
                  loadFavoriteOffers = {loadFavoriteOffers}
                  authorizationStatus = {authorizationStatus}
                  user = {user}
                  error = {error}
                />
              );
            }}
          />
          <Route
            render={() => {
              return <NotFound
                authorizationStatus = {authorizationStatus}
                user = {user}
                error = {error}
              />;
            }} />
        </Switch>
      </Router>
    );
  }
}

export const mapStateToProps = (state) => ({
  isFetching: state[NameSpace.DATA].isFetching,
  offers: getMemoizedOffers(state),
  cities: state[NameSpace.APP].cities,
  city: state[NameSpace.APP].city,
  cityOffers: getMemoizedCityOffers(state),
  error: state[NameSpace.DATA].error,
  authorizationStatus: state[NameSpace.AUTH].authorizationStatus,
  user: state[NameSpace.AUTH].user,
  favoriteOffers: state[NameSpace.DATA].favoriteOffers,
  reviews: state[NameSpace.DATA].reviews,
  nearbyOffers: state[NameSpace.DATA].nearbyOffers,
});

export const mapDispatchToProps = (dispatch) => ({
  login(authData) {
    dispatch(UserOperation.login(authData));
  },
  onMenuClick(city) {
    dispatch(ActionCreator.changeCity(city));
  },
  sendComment(comment, id) {
    dispatch(DataOperation.sendComment(comment, id))
      .then(() => dispatch(DataOperation.loadReviews(id)));
  },
  loadFavoriteOffers() {
    dispatch(DataOperation.loadFavoriteOffers());
  },
  loadReviews(id) {
    dispatch(DataOperation.loadReviews(id));
  },
  loadNearbyOffers(id) {
    dispatch(DataOperation.loadNearbyOffers(id));
  },
  onBookmarkButtonCLick(id, status) {
    dispatch(DataOperation.setToFavorite(id, status));
  },
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);

App.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.shape(placeCardType)).isRequired,
  cities: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  city: PropTypes.string.isRequired,
  cityOffers: PropTypes.arrayOf(PropTypes.shape(placeCardType)).isRequired,
  onMenuClick: PropTypes.func.isRequired,
  error: PropTypes.bool.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  user: PropTypes.string.isRequired,
  login: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  sendComment: PropTypes.func.isRequired,
  loadFavoriteOffers: PropTypes.func.isRequired,
  favoriteOffers: PropTypes.arrayOf(PropTypes.shape(placeCardType)).isRequired,
  loadReviews: PropTypes.func.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.shape(reviewType)).isRequired,
  nearbyOffers: PropTypes.arrayOf(PropTypes.shape(placeCardType)).isRequired,
  loadNearbyOffers: PropTypes.func.isRequired,
  onBookmarkButtonCLick: PropTypes.func.isRequired,
};
