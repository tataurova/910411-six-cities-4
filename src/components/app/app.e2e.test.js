import {mapStateToProps, mapDispatchToProps} from "./app.jsx";
import offers from "../../mocks/offers.js";
import {SortType} from "../../const.js";

describe(`<App />`, () => {
  it(`mapStateToProps returns initial state`, () => {
    const initialState = {
      city: `Paris`,
      offers: offers[`Paris`],
      sortType: SortType.DEFAULT,
      hoveredCardId: 0,
    };

    expect(mapStateToProps(initialState)).toEqual(initialState);
  });

  it(`mapDispatchToProps should be called for onSortTypeClick with right type and payload`, () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).onSortTypeClick(SortType.PRICE_UP);
    expect(dispatch.mock.calls[0][0]).toEqual({type: `CHANGE_SORT_TYPE`, payload: `price-up`});
  });

  it(`mapDispatchToProps should be called for onPlaceCardHover with right type and payload`, () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).onPlaceCardHover(1);
    expect(dispatch.mock.calls[0][0]).toEqual({type: `CHANGE_HOVERED_CARD`, payload: 1});
  });

  it(`mapDispatchToProps should be called for onMenuClick with right type and payload`, () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).onMenuClick(`Amsterdam`);
    expect(dispatch.mock.calls[0][0]).toEqual({type: `CHANGE_CITY`, payload: `Amsterdam`});
  });
});
