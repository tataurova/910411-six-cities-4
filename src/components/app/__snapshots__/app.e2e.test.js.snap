// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> App dispatch 1`] = `ShallowWrapper {}`;
